polkadot/node/core/approval-voting/src/lib.rs

```diff
@@ -1872,35 +1883,47 @@ async fn handle_from_overseer<Context>(
 			vec![Action::Conclude]
 		},
 		FromOrchestra::Communication { msg } => match msg {
-			ApprovalVotingMessage::CheckAndImportAssignment(a, claimed_cores, res) => {
+			ApprovalVotingMessage::CheckAndImportAssignment(a, claimed_cores, tranche, tx) => {
 				let (check_outcome, actions) = check_and_import_assignment(
 					ctx.sender(),
 					state,
 					db,
 					session_info_provider,
 					a,
 					claimed_cores,
+					tranche,
 				)
 				.await?;
-				let _ = res.send(check_outcome);
-
+				// approval-distribution already sanity checks the assignment is valid and expected,
+				// so this import should never fail if it does it means there is a bug in the
+				// code.
+				if let AssignmentCheckResult::Bad(ref err) = check_outcome {
+					gum::error!(target: LOG_TARGET, ?err, "Unexpected fail to check and import assignment");
+				}
+				let _ = tx.map(|tx| tx.send(check_outcome));
 				actions
 			},
-			ApprovalVotingMessage::CheckAndImportApproval(a, res) =>
-				check_and_import_approval(
+			ApprovalVotingMessage::CheckAndImportApproval(a, tx) => {
+				let result = check_and_import_approval(
 					ctx.sender(),
 					state,
 					db,
 					session_info_provider,
 					metrics,
 					a,
-					|r| {
-						let _ = res.send(r);
-					},
 					&wakeups,
 				)
-				.await?
-				.0,
+				.await?;
+				// approval-distribution already sanity checks the vote is valid and expected,
+				// so this import should never fail if it does it means there is a bug in the
+				// code.
+				if let ApprovalCheckResult::Bad(ref err) = result.1 {
```

Comment:
Sounds a bit like code smell. Is there an architecture in the design space that avoids those redundant checks?

Reply:
I think my comment is a bit misleading, the checks are not redundant, but this function can't be modified to not return an error because of two things:

1. It works with the IO/DB, there are paths where you might return error because those functions the can throw error.
2. Because of the parallelism inherent in our subsystem architecture, there is always the risk where 2 subsystems might get out of sync, so you always have to do basic sanity checks like is this block_hash known to the subsystem, is the candidate known to the subsystems, these are things you don't expect to fail but your better of validating them rather than assuming them. 

An error here signals probably a HW issue or a bug, so arguably we can panic here, but I wouldn't want to be so opinionated about this since it would make small bugs permanent, e.g: this problem: https://github.com/paritytech/polkadot-sdk/issues/5030, would actually make the node panic.

So, logging it and moving on seems like the right compromise in this situations.

Reply:
Can we just error out only for import failure and not checking the vote is valid?

Reply:
This function is not checking the vote is valid anymore.

I guess what I need to do is rename `check_and_import_approval` and `check_and_import_assignment`  into `import_approval` and `import_assignment` to avoid the confusion.

Reply:
Done.